---
title: "Pathway Analysis from RNA-Seq Results"
author: "Matt Hashimoto"
date: "11/29/2021"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Differential Expression Analysis

Lets first load the DESeq2 library:

```{r}
library(DESeq2)
```

Next, let's download and load our files into variables:

```{r}
# Establish filenames
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"
```

```{r}
# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)
```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

#### Q

"Complete the code below to remove the troublesome first column from countData."

```{r}
# Remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

The weird first column has been removed!

#### Q

"Complete the code below to filter countData to exclude genes (i.e. rows) where
we have 0 read count across all samples (i.e. columns)."

```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData[-which(rowSums(countData) == 0),]
head(countData)
```

## Running DESeq2

Let's set up the DESeqDataSet object:

```{r}
# Initialize the DESeqDataSet object
dds = DESeqDataSetFromMatrix(countData = countData,
                             colData = colData,
                             design =~ condition)
dds = DESeq(dds)

# View the object
dds
```

We can then take a look at the results:

```{r}
# Check results
res = results(dds, contrast = c("condition", "hoxa1_kd", "control_sirna"))
```

#### Q

"Call the summary() function on your results to get a sense of how many genes
are up or down-regulated at the default 0.1 p-value cutoff."

```{r}
summary(res)
```

## Volcano Plot

We can create a volcano plot to visually represent our data:

```{r}
# Use plot function to create a volcano plot
plot(res$log2FoldChange, -log(res$padj))
```

#### Q

"Improve this plot by completing the below code, which adds color and axis
labels."

```{r}
# Make a color vector for all genes
mycols <- rep("gray", nrow(res))

# Color red the genes with absolute fold change above 2
mycols[abs(res$log2FoldChange) > 2] <- "red"

# Color blue those with adjusted p-value less than 0.01 and absolute fold change
# more than 2
inds <- (res$padj < 0.01) & (abs(res$log2FoldChange) > 2)
mycols[inds] <- "blue"

plot(res$log2FoldChange, -log(res$padj), col = mycols,
     xlab = "Log2(FoldChange)", ylab = "-Log(P-value)" )
```

## Adding Gene Annotation

#### Q

"Use the mapIDs() function multiple times to add SYMBOL, ENTREZID and GENENAME
annotation to our results by completing the code below."

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

# Check possible column names
columns(org.Hs.eg.db)

# Add SYMBOL annotation
res$symbol = mapIds(org.Hs.eg.db,
                    keys = row.names(res),
                    keytype = "ENSEMBL",
                    column = "SYMBOL",
                    multiVals = "first")

# Add ENTREZID annotation
res$entrez = mapIds(org.Hs.eg.db,
                    keys = row.names(res),
                    keytype = "ENSEMBL",
                    column = "ENTREZID",
                    multiVals = "first")

# Add GENENAME annotation
res$name =   mapIds(org.Hs.eg.db,
                    keys = row.names(res),
                    keytype = "ENSEMBL",
                    column = "GENENAME",
                    multiVals = "first")

head(res, 10)
```

#### Q

"Finally for this section let's reorder these results by adjusted p-value and
save them to a CSV file in your current project directory."

```{r}
# Write results to a .csv file
res = res[order(res$pvalue),]
write.csv(res, file = "deseq_results.csv")
```

## Pathway Analysis

Now that we've downloaded the necessary packages, let's load them:

```{r}
library(pathview)
```

Let's next take a look at some of the available pathways:

```{r}
library(gage)
library(gageData)

data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

We can now go ahead and construct a vector with Entrez ID names for each index:

```{r}
# Create vector of fold changes, using Entrez IDs for names
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

We can now run gage analysis:

```{r}
# Get the results
keggres = gage(foldchanges, gsets = kegg.sets.hs)
```

For more details about the generated data:

```{r}
attributes(keggres)
```

Let's look at the first few "less" results:

```{r}
# Look at the first few down (less) pathways
head(keggres$less)
```

Let's view the pathway figure data by manually inputting the associated code:

```{r}
# View pathway data for hsa04110 Cell cycle
pathview(gene.data = foldchanges, pathway.id = "hsa04110")
```

```{r}
# A different PDF based output of the same data
pathview(gene.data = foldchanges, pathway.id = "hsa04110", kegg.native = FALSE)
```

We can use code to set up a way to automatically pull pathway codes from the
results from earlier:

```{r}
## Focus on top 5 upregulated pathways here for demo purposes only
keggrespathways <- rownames(keggres$greater)[1:5]

# Extract the 8 character long IDs part of each string
keggresids = substr(keggrespathways, start = 1, stop = 8)
keggresids
```

We can then use this vector to draw pathways for all:

```{r}
# View pathway data for top 5 results
pathview(gene.data = foldchanges, pathway.id = keggresids, species = "hsa")
```

#### Q

"Can you do the same procedure as above to plot the pathview figures for the top
5 down-reguled pathways?"

Yes, by looking at the data for "less" in the results instead of "greater":

```{r}
## Focus on top 5 downregulated pathways
keggrespathwaysDown <- rownames(keggres$less)[1:5]

# Extract the 8 character long IDs part of each string
keggresidsDown = substr(keggrespathwaysDown, start = 1, stop = 8)
keggresidsDown
```

```{r}
pathview(gene.data = foldchanges, pathway.id = keggresidsDown, species = "hsa")
```

## Gene Ontology

With a focus on Biological Process (BP), we can use gene ontology to analyze
the data in a similar way:

```{r}
data(go.sets.hs)
data(go.subs.hs)

# Focus on Biological Process subset of GO
gobpsets = go.sets.hs[go.subs.hs$BP]

gobpres = gage(foldchanges, gsets = gobpsets, same.dir = TRUE)

lapply(gobpres, head)
```

## Reactome Analysis

Before we can start analysis with Reactome, we must convert our gene data to
a plain .txt file:

```{r}
sig_genes <- res[res$padj <= 0.05 & !is.na(res$padj), "symbol"]
print(paste("Total number of significant genes:", length(sig_genes)))
```

```{r}
write.table(sig_genes, file="significant_genes.txt", row.names = FALSE,
            col.names = FALSE, quote = FALSE)
```

#### Q

"What pathway has the most significant “Entities p-value”? Do the most
significant pathways listed match your previous KEGG results? What factors
could cause differences between the two methods?"

The "Endosomal/Vascuolar pathway" has the most significant entities p-value.
The top results do differ from the most signficant pathways listed for the
previous KEGG results, but this may be due to the data being stored in different
ways for each database. Analysis is likely conducted differently between the two
as well. The Reactome pathways appear to be much more specific in terms of
pathways.

## GO Online

Gene Set GO Enrichment is another method of analysis.

#### Q

"What pathway has the most significant “Entities p-value”? Do the most
significant pathways listed match your previous KEGG results? What factors could
cause differences between the two methods?"

The "platelet-derived growth factor receptor signaling pathway" has the most
significant entities p-value. Again the most significant pathways are different
from the KEGG results. This may be due to differences between the storage of
data in terms of pathways, and the way genes are associated to each. There may
also be little consistency in the way pathways are self-contained.






























